pkg/telnetlib/connection.go

```diff
@@ -0,0 +1,345 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package telnetlib
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+type connOpts struct {
+	conn        net.Conn
+	fsm         *telnetFSM
+	cmdHandler  CmdHandlerFunc
+	dataHandler DataHandlerFunc
+	serverOpts  map[byte]bool
+	clientOpts  map[byte]bool
+	optCallback func(byte, byte)
+}
+
+// TelnetConn is the struct representing the telnet connection
+type TelnetConn struct {
+	// conn is the underlying connection
+	conn              net.Conn
+	readCh            chan []byte
+	writeCh           chan []byte
```

Comment:
The issue I guess I'm having (and it's really just curiosity) is why channels?  If there is a singleton  for reads and  another singleton for writes, why can't we just use io.Reader and io.Writer to/from the handlers?  Then you won't need the go routine below (connectionLoop) which basically implements io.Copy but for channels.